README
NAME
    hl7dump - Script to dump hl7 file message.

SYNOPSIS
     hl7dump [-c] [-h] [--version] hl7_file

ARGUMENTS
    *       "-c"

             Color mode.
             Color mode is controlled by ENV variables:
             - COLOR=1 - set color without -c argument.
             - NO_COLOR=1 - unset color with -c argument.

    *       "-h"

             Print help.

    *       "--version"

             Print version of script.

    *       "hl7_file"

             File with HL7 messages.

EXAMPLE
     hl7dump -h

REPOSITORY
    <https://github.com/michal-josef-spacek/App-HL7-Dump>

AUTHOR
    Michal Josef Špaček <mailto:skim@cpan.org>

    <http://skim.cz>

LICENSE AND COPYRIGHT
    © 2016-2023 Michal Josef Špaček

    BSD 2-Clause License

VERSION
    0.08



Changes
0.08

0.07 2023-05-11T18:20:29+02:00
 - Add better example names.
 - Fix repository author name.
 - Fix test on Windows.
 - Update copyright years.
 - Update Module::Install to 1.21 version.

0.06 2022-01-10T09:56:22+01:00
 - Fix test dependencies.

0.05 2022-01-06T14:48:31+01:00
 - Add test for constructor.
 - Add test for help.
 - Improve help message.
 - Move argument processing to run() method.
 - Regen example.
 - Update copyright years.

0.04 2021-06-28T15:16:30+02:00
 - Add exit code do run() method.
 - Add second example with color output.
 - Add support for COLOR and NO_COLOR env variables.
 - Fix minimal version of Perl.
 - Improve LICENSE AND COPYRIGHT secion in doc.
 - Improve SYNOPSIS secion in doc.
 - Rewrite doc to use =head2.
 - Update bugtracker to Github.
 - Update copyright years.

0.03 2018-02-07T08:06:09+01:00
 - Fix Makefile.PL for new version of Perl and no default »use lib '.'«.
 - Update Module::Install to 1.19 version.
 - Update author name.
 - Update copyright years.
 - Update distribution abstract.

0.02 2016-04-05T12:20:39+02:00
 - Add support for ANSI colors.
 - Fix ERRORS section in doc.
 - Fix SYNOPSIS section in doc.
 - Simplify of argument processing.

0.01 2016-04-03T22:02:04+02:00
 - First version.


Dump.pm
package App::HL7::Dump;

use strict;
use warnings;

use Class::Utils qw(set_params);
use English;
use Error::Pure qw(err);
use Getopt::Std;
use Net::HL7::Message;
use Perl6::Slurp qw(slurp);

our $VERSION = 0.08;

# Constructor.
sub new {
	my ($class, @params) = @_;

	# Create object.
	my $self = bless {}, $class;

	# Process params.
	set_params($self, @params);

	# Object.
	return $self;
}

# Run.
sub run {
	my $self = shift;

	# Process arguments.
	$self->{'_opts'} = {
		'c' => 0,
		'h' => 0,
	};
	if (! getopts('ch', $self->{'_opts'}) || @ARGV < 1
		|| $self->{'_opts'}->{'h'}) {

		print STDERR "Usage: $0 [-c] [-h] [--version] hl7_file\n";
		print STDERR "\t-c\t\tColor mode.\n";
		print STDERR "\t-h\t\tPrint help.\n";
		print STDERR "\t--version\tPrint version.\n";
		return 1;
	}
	$self->{'_hl7_file'} = $ARGV[0];

	if ($ENV{'COLOR'}) {
		$self->{'_opts'}->{'c'} = 1;
	}

	# Load Term::ANSIColor.
	if ($self->{'_opts'}->{'c'}) {
		eval "require Term::ANSIColor;";
		if ($EVAL_ERROR) {
			err "Cannot load 'Term::ANSIColor'.",
				'Eval error', $EVAL_ERROR;
		}
	}

	# Get hl7_file.
	my $hl7 = slurp($self->{'_hl7_file'});

	# Create message.
	my $msg = Net::HL7::Message->new($hl7);
	if (! $msg) {
		err 'Cannot parse HL7 file.', 'File', $self->{'_hl7_file'};
		return 1;
	}

	# Segment name: size
	foreach my $seg ($msg->getSegments) {
		foreach my $index (1 .. $seg->size) {
			my $val = $seg->getField($index);
			if (defined $val) {
				my $print_val;
				if (ref $val eq 'ARRAY') {
					$print_val = $seg->getFieldAsString($index);
				} else {
					$print_val = $val;
				}
				if ($self->{'_opts'}->{'c'}) {
					print Term::ANSIColor::color('green').$seg->getName.
						Term::ANSIColor::color('reset').'-'.
						Term::ANSIColor::color('red').$index.
						Term::ANSIColor::color('reset').':'.
						Term::ANSIColor::color('bold white').
						$print_val.Term::ANSIColor::color('reset')."\n";
				} else {
					print $seg->getName.'-'.$index.':'.$print_val."\n";
				}
			}
		}
	}

	return 0;
}

1;


__END__

=pod

=encoding utf8

=head1 NAME

App::HL7::Dump - Base class for hl7dump script.

=head1 SYNOPSIS

 use App::HL7::Dump;

 my $app = App::HL7::Dump->new;
 my $exit_code = $app->run;

=head1 METHODS

=head2 C<new>

 my $app = App::HL7::Dump->new;

Constructor.

Returns instance of object.

=head2 C<run>

 my $exit_code = $app->run;

Run method.

Returns exit code (0 as success, > 0 as error).

=head1 ERRORS

 new():
         Cannot load 'Term::ANSIColor'.
                 Eval error: %s
         From Class::Utils::set_params():
                 Unknown parameter '%s'.

 run():
         Cannot parse HL7 file.
                 File: %s

=head1 EXAMPLE1

=for comment filename=dump_hl7_message.pl

 use strict;
 use warnings;

 use App::HL7::Dump;
 use File::Temp qw(tempfile);
 use IO::Barf qw(barf);

 # Test data.
 my $hl7 = <<'END';
 MSH|^~\&|FROM|Facility #1|TO|Facility #2|20160403211012||ORM^O01|MSGID20160403211012|P|1.0
 PID|||11111||Novak^Jan^^^Ing.||19680821|M|||Olomoucká^^Brno^^61300^Czech Republic|||||||
 PV1||O|OP^PAREG^||||1234^Clark^Bob|||OP|||||||||2|||||||||||||||||||||||||20160403211012|
 ORC|NW|20160403211012
 OBR|1|20160403211012||003038^Urinalysis^L|||20160403211012
 END

 # Barf to temp file.
 my (undef, $file) = tempfile();
 barf($file, $hl7);

 # Arguments.
 @ARGV = (
         $file,
 );

 # Run.
 App::HL7::Dump->new->run;

 # Output:
 # MSH-1:|
 # MSH-2:^~\&
 # MSH-3:FROM
 # MSH-4:Facility #1
 # MSH-5:TO
 # MSH-6:Facility #2
 # MSH-7:20160403211012
 # MSH-9:ORM^O01
 # MSH-10:MSGID20160403211012
 # MSH-11:P
 # MSH-12:1.0
 # PID-3:11111
 # PID-5:Novak^Jan^^^Ing.
 # PID-7:19680821
 # PID-8:M
 # PID-11:Olomoucká^^Brno^^61300^Czech Republic
 # PV1-2:O
 # PV1-3:OP^PAREG
 # PV1-7:1234^Clark^Bob
 # PV1-10:OP
 # PV1-19:2
 # PV1-44:20160403211012
 # ORC-1:NW
 # ORC-2:20160403211012
 # OBR-1:1
 # OBR-2:20160403211012
 # OBR-4:003038^Urinalysis^L
 # OBR-7:20160403211012

=head1 EXAMPLE2

=for comment filename=dump_hl7_message_ansi.pl

 use strict;
 use warnings;

 use App::HL7::Dump;
 use File::Temp qw(tempfile);
 use IO::Barf qw(barf);

 # Test data.
 my $hl7 = <<'END';
 MSH|^~\&|FROM|Facility #1|TO|Facility #2|20160403211012||ORM^O01|MSGID20160403211012|P|1.0
 PID|||11111||Novak^Jan^^^Ing.||19680821|M|||Olomoucká^^Brno^^61300^Czech Republic|||||||
 PV1||O|OP^PAREG^||||1234^Clark^Bob|||OP|||||||||2|||||||||||||||||||||||||20160403211012|
 ORC|NW|20160403211012
 OBR|1|20160403211012||003038^Urinalysis^L|||20160403211012
 END

 # Barf to temp file.
 my (undef, $file) = tempfile();
 barf($file, $hl7);

 # Arguments.
 @ARGV = (
         '-c',
         $file,
 );

 # Run.
 App::HL7::Dump->new->run;

 # Output (colored keys):
 # MSH-1:|
 # MSH-2:^~\&
 # MSH-3:FROM
 # MSH-4:Facility #1
 # MSH-5:TO
 # MSH-6:Facility #2
 # MSH-7:20160403211012
 # MSH-9:ORM^O01
 # MSH-10:MSGID20160403211012
 # MSH-11:P
 # MSH-12:1.0
 # PID-3:11111
 # PID-5:Novak^Jan^^^Ing.
 # PID-7:19680821
 # PID-8:M
 # PID-11:Olomoucká^^Brno^^61300^Czech Republic
 # PV1-2:O
 # PV1-3:OP^PAREG
 # PV1-7:1234^Clark^Bob
 # PV1-10:OP
 # PV1-19:2
 # PV1-44:20160403211012
 # ORC-1:NW
 # ORC-2:20160403211012
 # OBR-1:1
 # OBR-2:20160403211012
 # OBR-4:003038^Urinalysis^L
 # OBR-7:20160403211012

=head1 DEPENDENCIES

L<Class::Utils>,
L<English>,
L<Error::Pure>,
L<Getopt::Std>,
L<Net::HL7::Message>,
L<Perl6::Slurp>.

L<Term::ANSIColor> for color mode.

=head1 REPOSITORY

L<https://github.com/michal-josef-spacek/App-HL7-Dump>

=head1 AUTHOR

Michal Josef Špaček L<mailto:skim@cpan.org>

L<http://skim.cz>

=head1 LICENSE AND COPYRIGHT

© 2016-2023 Michal Josef Špaček

BSD 2-Clause License

=head1 VERSION

0.08

=cut


Makefile.PL
use lib '.';
use strict;
use warnings;

use inc::Module::Install;

# Definition.
abstract 'Distribution for hl7dump script.';
author 'Michal Josef Spacek <skim@cpan.org>';
author_requires 'English' => 0;
author_requires 'File::Object' => 0.05;
author_requires 'Test::More' => 0;
author_requires 'Test::NoWarnings' => 0;
author_requires 'Test::Pod' => 0;
author_requires 'Test::Pod::Coverage' => 0;
install_script 'bin/hl7dump';
license 'bsd';
name 'App-HL7-Dump';
perl_version 5.006;
readme_from 'bin/hl7dump';
recommends 'Term::ANSIColor' => 5.01;
recursive_author_tests('xt');
requires 'Class::Utils' => 0.03;
requires 'English' => 0;
requires 'Error::Pure' => 0.15;
requires 'Getopt::Std' => 0;
requires 'Net::HL7::Message' => 0;
requires 'Perl6::Slurp' => 0;
resources 'bugtracker' => 'https://github.com/michal-josef-spacek/App-HL7-Dump/issues';
resources 'homepage' => 'https://github.com/michal-josef-spacek/App-HL7-Dump';
resources 'repository' => 'git://github.com/michal-josef-spacek/App-HL7-Dump';
test_requires 'English' => 0;
test_requires 'File::Object' => 0;
test_requires 'File::Spec::Functions' => 0;
test_requires 'Test::More' => 0;
test_requires 'Test::NoWarnings' => 0;
test_requires 'Test::Output' => 0;
tests_recursive;
version '0.08';

# Run.
WriteAll();
